dask_cuda/device_host_file.py

```diff
@@ -54,43 +48,30 @@ def _(obj):
         headers.append(header)
         all_frames.extend(frames)
 
-    header = {"sub-headers": headers, "is-cuda": obj.is_cuda, "main-header": obj.header}
+    header = {"sub-headers": headers, "main-header": obj.header}
 
     return header, all_frames
 
 
 @dask_deserialize.register(DeviceSerialized)
-def _(header, frames):
+def device_deserialize(header, frames):
     parts = []
     for sub_header in header["sub-headers"]:
         start, stop = sub_header.pop("frame-start-stop")
         part = deserialize(sub_header, frames[start:stop])
         parts.append(part)
 
-    return DeviceSerialized(header["main-header"], parts, header["is-cuda"])
-
-
-def copy_to_host(ary):
-    if hasattr(ary, "copy_to_host"):
-        return ary.copy_to_host()
-    else:
-        return cuda.as_cuda_array(ary).copy_to_host()
+    return DeviceSerialized(header["main-header"], parts)
 
 
 def device_to_host(obj: object) -> DeviceSerialized:
-    header, frames = serialize(obj, serializers=["cuda", "pickle"])
-    is_cuda = [hasattr(f, "__cuda_array_interface__") for f in frames]
-    frames = [copy_to_host(f) if ic else f for ic, f in zip(is_cuda, frames)]
-    return DeviceSerialized(header, frames, is_cuda)
+    header, frames = serialize(obj, serializers=["dask", "pickle"])
+    frames = [numpy.asarray(f) for f in frames]
```

Comment:
Just a note here, the `frames` are `memoryviews`. Since we want to serialize `DeviceSerialized` later, which involves serializing these `frames`, and Dask doesn't know how to serialize a `memoryview`, it fails. To fix this, we construct a NumPy `ndarray` that merely views the underlying data in the `memoryview`. IOW this quite fast and zero-copy. Since Dask knows how to serialize NumPy `ndarray`s, this should proceed without issues.

Reply:
Got it, this makes sense. Thanks for the description of how it works, it was really helpful to make it understand what's going on in here.

Reply:
Yeah ideally we wouldn't need this, but for now its a simple enough workaround 🙂